[PodSecurity] baseline - apparmor

n4j opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Implement the "AppArmor" check from https://kubernetes.io/docs/concepts/security/pod-security-standards/#baseline
- Disallows `containers` and `initContainers` with `uncofined` AppArmor annotation

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubernetes/issues/103200

#### Special notes for your reviewer:
NONE
#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/2579-psp-replacement
```

n4j commented:
/assign @tallclair @liggitt

liggitt commented:
@tallclair will take lead reviewing this one

liggitt commented:
/ok-to-test

tallclair commented on staging/src/k8s.io/pod-security-admission/policy/check_apparmor.go:
```diff
@@ -0,0 +1,93 @@
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/pod-security-admission/api"
+)
+
```
nit: can you past the AppArmor standards description (from https://kubernetes.io/docs/concepts/security/pod-security-standards/#baseline) into a comment here (e.g. https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/pod-security-admission/policy/check_allowPrivilegeEscalation.go#L28-L37)

tallclair commented on staging/src/k8s.io/pod-security-admission/policy/check_apparmor.go:
```diff
@@ -0,0 +1,93 @@
+package policy
+
+import (
+	"fmt"
```
nit: line break after stdlib imports (`make update` might fix this automatically, not sure)

tallclair commented on staging/src/k8s.io/pod-security-admission/policy/check_apparmor.go:
```diff
@@ -0,0 +1,93 @@
+	"fmt"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
```
nit: delete this line break

tallclair commented on staging/src/k8s.io/pod-security-admission/policy/check_apparmor.go:
```diff
@@ -0,0 +1,93 @@
+}
+
+const (
+	permissionFormat = "container.apparmor.security.beta.kubernetes.io/%s"
```
use `corev1.AppArmorBetaContainerAnnotationKeyPrefix`

tallclair commented on staging/src/k8s.io/pod-security-admission/policy/check_apparmor.go:
```diff
@@ -0,0 +1,93 @@
+var (
+	forbiddenValues = sets.NewString(
+		"unconfined",
+	)
```
I'd prefer to approach this as an allowlist instead. e.g. 

```go
func allowedProfile(profile string) {
  return profile == corev1.AppArmorBetaProfileRuntimeDefault || strings.HasPrefix(profile, corev1.AppArmorBetaProfileNamePrefix)
}
```

tallclair commented on staging/src/k8s.io/pod-security-admission/policy/check_apparmor.go:
```diff
@@ -0,0 +1,93 @@
+	}
+
+	// check containers
+	for _, container := range podSpec.Containers {
```
Rather than checking each container, I think it would be better to just check each annotation that matches the AppArmor prefix. It doesn't matter if there isn't an associated container, it should still fail.

tallclair commented on staging/src/k8s.io/pod-security-admission/test/fixtures_apparmor.go:
```diff
@@ -0,0 +1,123 @@
+	permissionFormat = "container.apparmor.security.beta.kubernetes.io/%s"
+)
+
+func ensureAnnotation(pod *corev1.Pod) *corev1.Pod {
```
nit: move to `helpers.go`

tallclair reviewed (changes requested):
Thanks for picking this up!

n4j commented:
@tallclair Thanks for the PR comments :)

Can you please re-review?

tallclair commented on staging/src/k8s.io/pod-security-admission/policy/check_apparmor.go:
```diff
@@ -0,0 +1,88 @@
+	}
+
+	for k, v := range podMetadata.Annotations {
+		if strings.HasPrefix(k, corev1.AppArmorBetaContainerAnnotationKeyPrefix); !allowedProfile(v) {
```
```suggestion
		if strings.HasPrefix(k, corev1.AppArmorBetaContainerAnnotationKeyPrefix) && !allowedProfile(v) {
```

I don't think the initializer statement is evaluated in an if statement (although I'm surprised the tests pass, so maybe it does?)

liggitt replied:
it does, and it doesn't do what you expect: https://play.golang.org/p/k7zxAkljE2o

this check will fail if any annotation has any value other than `runtime/default` and does not start with `localhost/`, regardless of the annotation key

tallclair commented on staging/src/k8s.io/pod-security-admission/test/fixtures_apparmor.go:
```diff
@@ -0,0 +1,111 @@
+
+				// container with localhost/foo annotation
+				tweak(pod, func(copy *corev1.Pod) {
+					name := "test"
+					if len(copy.Spec.Containers) > 0 {
+						copy.Spec.Containers[0].Name = name
+					}
```
rather than setting the name, just use the existing name. You can also skip the length check, so this just becomes:
```suggestion
					name := copy.Spec.Containers[0].Name
```

tallclair replied:
Same for other tests (init containers could technically be empty, but this is just in a test so we can depend on the minimalViablePod having an init container)

tallclair commented on staging/src/k8s.io/pod-security-admission/test/fixtures_apparmor.go:
```diff
@@ -0,0 +1,111 @@
+					if len(copy.Spec.Containers) > 0 {
+						copy.Spec.Containers[0].Name = name
+					}
+				}),
```
You can skip this test. It's covered by the minimum viable pod.

tallclair commented on staging/src/k8s.io/pod-security-admission/test/fixtures_apparmor.go:
```diff
@@ -0,0 +1,111 @@
+					if len(copy.Spec.InitContainers) > 0 {
+						copy.Spec.InitContainers[0].Name = name
+					}
+				}),
```
same as above, drop this case.

tallclair commented on staging/src/k8s.io/pod-security-admission/test/fixtures_apparmor.go:
```diff
@@ -0,0 +1,111 @@
+					if len(copy.Spec.InitContainers) > 0 {
+						copy.Spec.InitContainers[0].Name = name
+					}
+				}),
```
I think you can drop this case. 1 test case for initContainers should be enough.

liggitt commented on staging/src/k8s.io/pod-security-admission/test/fixtures_apparmor.go:
```diff
@@ -0,0 +1,111 @@
+func init() {
+	appArmorFixture_1_0 := fixtureGenerator{
+		expectErrorSubstring: "forbidden AppArmor profile",
+		generatePass: func(pod *corev1.Pod) []*corev1.Pod {
```
make sure the passing test fixture (or a unit test) exercises treatment of non-app-armor annotations

n4j commented:
/sig auth

n4j commented:
@tallclair Can you please re-review?

liggitt commented on staging/src/k8s.io/pod-security-admission/test/fixtures_apparmor.go:
```diff
@@ -0,0 +1,73 @@
+		generatePass: func(pod *corev1.Pod) []*corev1.Pod {
+			pod = ensureAnnotation(pod)
+			return []*corev1.Pod{
+				// container with runtime/default annotation
+				tweak(pod, func(copy *corev1.Pod) {
+					name := copy.Spec.Containers[0].Name
+					copy.Annotations[corev1.AppArmorBetaContainerAnnotationKeyPrefix+name] = "runtime/default"
+					copy.Annotations["foo"] = "bar"
+				}),
+
+				// container with localhost/foo annotation
+				tweak(pod, func(copy *corev1.Pod) {
+					name := copy.Spec.Containers[0].Name
+					copy.Annotations[corev1.AppArmorBetaContainerAnnotationKeyPrefix+name] = "localhost/foo"
+					copy.Annotations["foo"] = "bar"
+				}),
+
+				// initContainer with localhost/foo annotation
+				tweak(pod, func(copy *corev1.Pod) {
+					name := copy.Spec.Containers[0].Name
+					copy.Annotations[corev1.AppArmorBetaContainerAnnotationKeyPrefix+name] = "localhost/foo"
+				}),
```
combine these into a single passing testcase that sets multiple annotations... that reduces fixture count without dropping coverage

also, drop the foo=bar annotation from the fixtures... we'll test things like that in check unit tests

liggitt commented on staging/src/k8s.io/pod-security-admission/test/fixtures_apparmor.go:
```diff
@@ -0,0 +1,73 @@
+
+				// initContainer with unconfined annotation
+				tweak(pod, func(copy *corev1.Pod) {
+					name := copy.Spec.Containers[0].Name
```
should be the initContainer name

liggitt commented:
looks good, just a couple tweaks to the test fixtures (don't forget to regenerate fixtures and squash after making the updates)

n4j commented:
/test pull-kubernetes-integration

liggitt commented on pkg/security/podsecuritypolicy/sysctl/mustmatchpatterns.go:
```diff
@@ -35,6 +35,7 @@ func SafeSysctlWhitelist() []string {
 		"net.ipv4.ip_local_port_range",
 		"net.ipv4.tcp_syncookies",
 		"net.ipv4.ping_group_range",
+		"net.ipv4.ip_unprivileged_port_start",
```
I'm surprised to see this... bad rebase?

n4j replied:
Yes, give me a moment to fix this

n4j committed:
[PodSecurity] baseline - apparmor
Implement the "AppArmor" check from https://kubernetes.io/docs/concepts/security/pod-security-standards/#baseline

- AppArmor check
- Fixtures
- UnitTest case

n4j commented:
Raised https://github.com/kubernetes/kubernetes/pull/103412 due to bad rebase issues

tallclair commented:
@n4j I'd prefer to keep this PR open to maintain the comment history. Ping me on slack if you want, and i can help you resolve the rebase issues. If https://github.com/kubernetes/kubernetes/pull/103412 has what you want, the following should work:

```sh
git checkout feature/podSecurityApparmor_v2
git reset --hard feature/appArmor_v3
git push -f origin feature/podSecurityApparmor_v2
```

n4j commented:
@tallclair / @liggitt Please re-review.

tallclair commented on staging/src/k8s.io/pod-security-admission/policy/check_apparmor_test.go:
```diff
@@ -0,0 +1,80 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestCheckAppArmor(t *testing.T) {
```
thanks for the unit test!

tallclair reviewed (approved):
/lgtm
/approve

Thanks!
